src/app/content/highlights/components/Highlights.tsx

```diff
@@ -0,0 +1,128 @@
+  const highlights = useSelector(summaryHighlights);
+  const isLoading = useSelector(summaryIsLoading);
+
+  if (
+    !isLoading &&
+    (filters.colors.length === 0 || filters.locationIds.length === 0)
+  ) {
+    return <Styled.Highlights>
+      <HStyled.GeneralCenterText>
+        <FormattedMessage id='i18n:toolbar:highlights:popup:heading:no-highlights'>
+          {(msg: Element | string) => msg}
+        </FormattedMessage>
+        <span>
+          <FormattedMessage
+            id='i18n:toolbar:highlights:popup:heading:no-highlights-tip'
+            defaultMessage='Try selecting different chapter or color filters to see different results.'
+            values={{strong: (str: string) => <strong>{str}</strong>}}
+          />
+        </span>
+      </HStyled.GeneralCenterText>
+    </Styled.Highlights>;
+  }
+
+  if (
+    isLoading ||
+    (locations.size > 0 && Object.keys(highlights).length > 0)
+  ) {
+    return <Styled.Highlights isLoading={isLoading}>
+      {isLoading ? <Styled.LoaderWrapper><Loader/></Styled.LoaderWrapper> : null}
+      {Array.from(locations).map(([id, location]) => {
+        if (!highlights[id]) { return null; }
+        return <SectionHighlights
+          key={id}
+          location={location}
+          highlights={highlights}
+        />;
+      })}
+    </Styled.Highlights>;
+  }
+
+  return <Styled.Highlights>
+    <HStyled.GeneralLeftText>
+      <FormattedMessage id='i18n:toolbar:highlights:popup:body:no-highlights-in-chapter'>
+        {(msg: Element | string) => msg}
+      </FormattedMessage>
+    </HStyled.GeneralLeftText>
+    <HStyled.MyHighlightsWrapper>
+      <HStyled.GeneralText>
+        <FormattedMessage id='i18n:toolbar:highlights:popup:body:add-highlight'>
+          {(msg: Element | string) => msg}
+        </FormattedMessage>
+      </HStyled.GeneralText>
+      <HStyled.GeneralTextWrapper>
+        <FormattedMessage id='i18n:toolbar:highlights:popup:body:use-this-page'>
+          {(msg: Element | string) => msg}
+        </FormattedMessage>
+      </HStyled.GeneralTextWrapper>
+      <HStyled.MyHighlightsImage src={myHighlightsEmptyImage} />
+    </HStyled.MyHighlightsWrapper>
+  </Styled.Highlights>;
+};
+
+export default Highlights;
+
+interface SectionHighlightsProps {
+  location: LinkedArchiveTreeNode;
+  highlights: SummaryHighlights;
+}
+
+// tslint:disable-next-line: variable-name
+export const SectionHighlights = ({ location, highlights }: SectionHighlightsProps) => {
+  const pageIdIsSameAsSectionId = highlights[location.id][location.id];
+  return (
+    <React.Fragment>
+      {!pageIdIsSameAsSectionId && <Styled.HighlightsChapter
+        dangerouslySetInnerHTML={{ __html: location.title }}
+      />}
+      {Object.entries(highlights[location.id]).map(([pageId, pageHighlights]) => {
+        pageId = stripIdVersion(pageId);
+        const page = archiveTreeSectionIsChapter(location)
+          ? findArchiveTreeNode(location, pageId)!
```

Comment:
Should I in this case write separate test case to handle this branch?

Reply:
if something really _should_ never happen, its ok to throw an exception (for example by using the assertDefined utility), that makes testing easier because it doesn't create a branch that needs coverage. 

if it is reasonably possible for the null case to actually occur, then yea, you'd need to add a short circuit and add a test case for that branch.